### PR TITLE
[SPARK-21981][PYTHON][ML] Added Python interface for ClusteringEvaluator

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Added Python interface for ClusteringEvaluator

## How was this patch tested?

Manual test, eg. the example Python code in the comments.

cc @yanboliang

### CODE REVIEW

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

**Review by WeichenXu123 — COMMENTED** [review]
We'd better add metrics interface for the python api, (including metricName Param, setMetricName/getMetricName method), although there is only one metrics for now.

**Comment by mgaido91** [comment]
Thanks @WeichenXu123, I added it.

**Comment by gatorsmile** [comment]
ok to test

File: python/pyspark/ml/evaluation.py [hunk 1]
```diff
@@ -328,6 +329,87 @@ def setParams(self, predictionCol="prediction", labelCol="label",
...
+    >>> evaluator.save(ce_path)
+    >>> evaluator2 = ClusteringEvaluator.load(ce_path)
+    >>> str(evaluator2.getPredictionCol())
+    'cluster_id'
+
+    .. versionadded:: 2.3.0
+    """
+    metricName = Param(Params._dummy(), "metricName",
+                       "metric name in evaluation "
+                       "(silhouette)",
```
Comment (WeichenXu123) [inline python/pyspark/ml/evaluation.py]: The string in multiple lines, we should use """ instead of "".  Otherwise move them to the same line.

**Review by WeichenXu123 — COMMENTED** [review]
LGTM except one comment. Thanks!

**Review by WeichenXu123 — APPROVED** [review]
LGTM. Thanks!

**Comment by mgaido91** [comment]
Thank you for your review and help @WeichenXu123!

File: python/pyspark/ml/evaluation.py [hunk 2]
```diff
@@ -328,6 +329,86 @@ def setParams(self, predictionCol="prediction", labelCol="label",
...
+
+    Evaluator for Clustering results, which expects two input
+    columns: prediction and features.
+
+    >>> from sklearn import datasets
+    >>> from pyspark.sql.types import *
+    >>> from pyspark.ml.linalg import Vectors, VectorUDT
+    >>> from pyspark.ml.evaluation import ClusteringEvaluator
+    ...
+    >>> iris = datasets.load_iris()
```
Comment (yanboliang) [inline python/pyspark/ml/evaluation.py]: Please don't involves other libraries if not necessary, here the doc test is used to show how to use ```ClusteringEvaluator```  to fresh users, so we should focus on evaluator and keep it as simple as possible. You can refer other evaluator to construct simple dataset.

File: python/pyspark/ml/evaluation.py [hunk 3]
```diff
@@ -328,6 +329,86 @@ def setParams(self, predictionCol="prediction", labelCol="label",
...
+    """
+    .. note:: Experimental
+
+    Evaluator for Clustering results, which expects two input
+    columns: prediction and features.
+
+    >>> from sklearn import datasets
+    >>> from pyspark.sql.types import *
+    >>> from pyspark.ml.linalg import Vectors, VectorUDT
+    >>> from pyspark.ml.evaluation import ClusteringEvaluator
```
Comment (yanboliang) [inline python/pyspark/ml/evaluation.py]: Remove this, it's not necessary.

File: python/pyspark/ml/evaluation.py [hunk 4]
```diff
@@ -328,6 +329,86 @@ def setParams(self, predictionCol="prediction", labelCol="label",
...
+    def __init__(self, predictionCol="prediction", featuresCol="features",
+                 metricName="silhouette"):
+        """
+        __init__(self, predictionCol="prediction", featuresCol="features", \
+                 metricName="silhouette")
+        """
+        super(ClusteringEvaluator, self).__init__()
+        self._java_obj = self._new_java_obj(
+            "org.apache.spark.ml.evaluation.ClusteringEvaluator", self.uid)
+        self._setDefault(predictionCol="prediction", featuresCol="features",
```
Comment (yanboliang) [inline python/pyspark/ml/evaluation.py]: Remove setting default value for ```predictionCol``` and ```featuresCol```, as they have been set in ```HasPredictionCol``` and ```HasFeaturesCol```.

File: python/pyspark/ml/evaluation.py [hunk 5]
```diff
@@ -328,6 +329,86 @@ def setParams(self, predictionCol="prediction", labelCol="label",
...
+    >>> from pyspark.sql.types import *
+    >>> from pyspark.ml.linalg import Vectors, VectorUDT
+    >>> from pyspark.ml.evaluation import ClusteringEvaluator
+    ...
+    >>> iris = datasets.load_iris()
+    >>> iris_rows = [(Vectors.dense(x), int(iris.target[i]))
+    ...     for i, x in enumerate(iris.data)]
+    >>> schema = StructType([
+    ...    StructField("features", VectorUDT(), True),
+    ...    StructField("cluster_id", IntegerType(), True)])
```
Comment (yanboliang) [inline python/pyspark/ml/evaluation.py]: ```cluster_id``` -> ```prediction``` to emphasize this is the prediction value, not ground truth.

File: python/pyspark/ml/evaluation.py (hunk 4 above)
Comment (yanboliang) [reply python/pyspark/ml/evaluation.py]: I sent #19262 to fix same issue for other evaluators, please feel free to comment. Thanks.

File: python/pyspark/ml/evaluation.py [hunk 6]
```diff
@@ -328,6 +329,77 @@ def setParams(self, predictionCol="prediction", labelCol="label",
...
+class ClusteringEvaluator(JavaEvaluator, HasPredictionCol, HasFeaturesCol,
+                          JavaMLReadable, JavaMLWritable):
+    """
+    .. note:: Experimental
+
+    Evaluator for Clustering results, which expects two input
+    columns: prediction and features.
+
+    >>> from pyspark.ml.linalg import Vectors
+    >>> scoreAndLabels = map(lambda x: (Vectors.dense(x[0]), x[1]),
```
Comment (yanboliang) [inline python/pyspark/ml/evaluation.py]: ```scoreAndLabels``` -> ```featureAndPredictions```, the dataset here is different from other evaluators, we should use more accurate name. Thanks.

**Review by yanboliang — COMMENTED** [review]
One minor issue, otherwise, LGTM. Thanks.

**Comment by yanboliang** [comment]
Merged into master, thanks.

**Pull request closed without merging**